bpf: lb: deal with stale rev_nat_index after svc lookup in fallback path

julianwiedmann opened the pull request:
Fix two corner-case bugs in the fallback path for `lb_local()`, when a previously selected backend is unavailable. In this case we perform a fresh svc lookup, and thus need to handle that the new svc entry is associated with a different `rev_nat_index`.

julianwiedmann committed:
bpf: nodeport: fix rev_nat_index check for stale NAT CT entries

For non-DSR connections, the nodeport code creates a special CT_EGRESS
entry to maintain its own state (and eg. trigger RevDNAT from it).

This CT entry also tracks the service's rev_nat_index. On a successful
CT lookup, we detect a stale entry by comparing its rev_nat_index to
the currently matched svc's rev_nat_index.

But there's a special case in lb*_lookup() where the backend lookup fails,
and we fall back to a fresh svc lookup. The new svc entry isn't reflected
back to the callers, and consequently they would use the wrong
svc->rev_nat_index to check for staleness of their CT entry.

The actual rev_nat_index *is* reflected back to the caller via the
passed-in ct_state_new struct. So use that for the staleness check instead.

Note that this matches how bpf_lxc determines the current rev_nat_index
in lb*_ctx_store_state().

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: lb: update revNAT index after SVC change

lb*_local() has a fallback path for when the previously selected backend
for a svc connection is no longer available. In this case we perform a
fresh svc lookup, and select a new backend.

As this can potentially return a different svc entry, make sure to also
update the rev_nat_index in the CT_SERVICE entry. Otherwise the backend
lookup for the *next* packet will detect a mismatch between svc and
CT entry, and trigger a fresh backend selection.

Also reflect the new rev_nat_index back to the caller in the ct_state
struct.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

gentoo-root commented on bpf/lib/nodeport.h:
```diff
@@ -1028,7 +1028,7 @@ static __always_inline int nodeport_lb6(struct __ctx_buff *ctx,
 		case CT_ESTABLISHED:
 			if (unlikely(ct_state.rev_nat_index !=
-				     svc->rev_nat_index))
+				     ct_state_new.rev_nat_index))
```
> The new svc entry isn't reflected back to the callers

But it not only affects rev_nat_index. nodeport_lb4 continues using svc after lb4_local returns. Isn't it problematic too?

julianwiedmann replied:
yep. Not trying to fix the world in one day - making this part consistent with bpf_lxc is progress already :)

gentoo-root replied:
I agree it's still progress, but passing svc up to the caller doesn't seem much harder, and it makes this change unnecessary while fixing more bugs. Up to you, though; but let's have an issue to track the other bugs if you are not fixing them in this pull request.

julianwiedmann replied:
>but passing svc up to the caller doesn't seem much harder

It's not that trivial I believe. We've already done checks on the svc *before* calling lb4_local(), so would need to duplicate / reorder those.

The nicer fix would be to understand why we even need that fresh svc lookup, and figure out if we can do this a bit more graceful (reset the `backend_id` in the CT entry, drop the packet, let the calling context do the fresh svc lookup on the retransmit?). But -ENOCYCLES, this fix has been dangling for too long already :).

>but let's have an issue to track the other bugs if you are not fixing them in this pull request.

:+1:

gentoo-root replied:
OK, I see, if it's not trivial, let's just track it then.

julianwiedmann replied:
-> https://github.com/cilium/cilium/issues/25373

gentoo-root commented on bpf/lib/lb.h:
```diff
@@ -935,6 +935,7 @@ static __always_inline int lb6_local(const void *map, struct __ctx_buff *ctx,
 			goto drop_no_service;
 		state->backend_id = backend_id;
 		ct_update_backend_id(map, tuple, state);
+		ct_update_rev_nat_index(map, tuple, svc->rev_nat_index);
```
One block above we have:

```
		state->backend_id = backend_id;
		ct_update_backend_id(map, tuple, state);
		state->rev_nat_index = svc->rev_nat_index;
		ct_update_rev_nat_index(map, tuple, state);
```

I struggle to understand the motivation why you replaced state with raw rev_nat_index in the previous commit (this should be equivalent, because we assign `state->rev_nat_index = svc->rev_nat_index` a few lines below anyway), at the same time kept two similar snippets inconsistent. Also, both ct_update_backend_id and ct_update_rev_nat_index seem to be used always together, so how about we optimize out one map_lookup_elem, combine these two functions into one and just pass the new updated state?

```
		state->backend_id = backend_id;
		state->rev_nat_index = svc->rev_nat_index;
		ct_update_svc_state(map, tuple, state); /* Whatever name you come up with. */
```

(Or maybe even put the assignments inside that function to further reduce repeating code.)

It seems that then we can drop `state->rev_nat_index = svc->rev_nat_index` under update_state altogether, because it will be a no-op in all flows.

--

Also a little bit suspicious that we don't check `!backend_id` in this block, unlike the previous one. Is it redundant for some reason? I'm just thinking in the direction of further deduplicating these two blocks.

julianwiedmann replied:
> Also, both ct_update_backend_id and ct_update_rev_nat_index seem to be used always together, so how about we optimize out one map_lookup_elem, combine these two functions into one and just pass the new updated state?

ack, that's a nice follow-on cleanup! Will queue up for the next PR.

julianwiedmann replied:
>I struggle to understand the motivation why you replaced state with raw rev_nat_index in the previous commit (this should be equivalent, because we assign state->rev_nat_index = svc->rev_nat_index a few lines below anyway), at the same time kept two similar snippets inconsistent. 

As you noted, the `state->rev_nat_index = svc->rev_nat_index` updates are mostly redundant - as we update it either way in the `update_state` epilogue. So I can obviously still add it along with the new `ct_update_rev_nat_index()` to keep things consistent, but didn't want to add unnecessary code. Any preference?

julianwiedmann replied:
>Also a little bit suspicious that we don't check !backend_id in this block, unlike the previous one. Is it redundant for some reason? I'm just thinking in the direction of further deduplicating these two blocks.

yeah that's why I'm slowly chipping away at the whole code path - to get a better picture of what other bugs are still lurking. Some of the `!backend_id` checks are just to respect session-affinity (which we don't want in this fallback path) - but I agree, that doesn't explain everything.

gentoo-root replied:
> So I can obviously still add it along with the new ct_update_rev_nat_index() to keep things consistent, but didn't want to add unnecessary code. Any preference?

I think, you can remove the assignment under the update_state label, we jump there from CT_NEW where this assignment was already made, and from CT_REPLY where we also always end up with equal values (as long as you add it with the new ct_update_rev_nat_index).

This way you won't have redundant code (in fact, even remove redundancy on CT_NEW), and it's still consistent with the block above. Unless I missed something, that sounds pretty clean.

julianwiedmann replied:
>I think, you can remove the assignment under the update_state label, we jump there from CT_NEW where this assignment was already made, and from CT_REPLY where we also always end up with equal values (as long as you add it with the new ct_update_rev_nat_index).

Huh. The calling context expects to always find a valid `ct_state->rev_nat_index` - so I think removing the assignment in `update_state` will be slightly more involved. But sure, let's try.

gentoo-root replied:
Did I miss any flow? Because it seems that it will be functionally equivalent in all cases. We either set `state->rev_nat_index = svc->rev_nat_index` explicitly, or check that they are already equal, in which case setting it again is not needed. The caller will have the valid ct_state_new with rev_nat_index set at all times, unless we early return on error.

I don't understand the concern. If I missed something, please elaborate.

julianwiedmann replied:
Note how the `CT_REPLY` case will typically *not* do the assignment (it checks for `state->rev_nat_index == 0`, not for eg. `state->rev_nat_index == svc->rev_nat_index`). So in its common path, it currently relies on the assignment at `update_state`.

julianwiedmann replied:
nevermind, I see what you mean. Too much spaghetti :)

gentoo-root replied:
> Note how the CT_REPLY case will typically not do the assignment

That's fine, because it exits the switch, and the first check after the switch is `if (state->rev_nat_index != svc->rev_nat_index) {`.

julianwiedmann replied:
>ack, that's a nice follow-on cleanup! Will queue up for the next PR.

stashed in https://github.com/cilium/cilium/pull/25372 :heart:

julianwiedmann committed:
bpf: lb: clean up state->rev_nat_index assignments in lb*_local()

The caller expects us to reflect back a valid state->rev_nat_index
(that matches svc->rev_nat_index).

- On CT_NEW we already assign this, as part of creating the CT entry.
- On CT_REPLY the obtained state->rev_nat_index will already contain some
  value, and we subsequently check whether it still matches the selected
  svc entry (and update accordingly).

So the additional assignment at the `update_state` label is not required.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>
